src/main/java/net/glowstone/block/itemtype/ItemChorusFruit.java

```diff
@@ -0,0 +1,74 @@
+package net.glowstone.block.itemtype;
+
+import net.glowstone.entity.GlowPlayer;
+import org.bukkit.Location;
+import org.bukkit.Sound;
+import org.bukkit.SoundCategory;
+import org.bukkit.World;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+import org.bukkit.inventory.ItemStack;
+
+public class ItemChorusFruit extends ItemFood {
+
+    public ItemChorusFruit() {
+        super(4, 2.4f);
+    }
+
+    @Override
+    public boolean eat(GlowPlayer player, ItemStack item) {
+        if (!super.eat(player, item)) {
+            return false;
+        }
+
+        for (int i = 0; i < 16; i++) { //16 attempts: +/- 8 blocks in every direction
+            Location attempt = player.getLocation();
+            double deltaX = Math.random() * 16 - 8;
+            double deltaY = Math.random() * 16 - 8;
+            double deltaZ = Math.random() * 16 - 8;
+            attempt.setX(attempt.getX() + deltaX);
+            attempt.setY(Math.min(Math.max(attempt.getY() + deltaY, 0), player.getWorld().getMaxHeight() - 1));
+            attempt.setZ(attempt.getZ() + deltaZ);
+            attempt = getSafeLocation(attempt);
+            if (attempt != null) {
+                player.getWorld().playSound(player.getLocation(), Sound.ITEM_CHORUS_FRUIT_TELEPORT, SoundCategory.PLAYERS, 1, 1);
+                player.teleport(attempt);
+                break;
+            }
+        }
+        return true;
+    }
+
+    private Location getSafeLocation(Location loc) {
+        int yCoord = loc.getBlockY();
+        World world = loc.getWorld();
+        if (yCoord > world.getHighestBlockYAt(loc)) {
+            yCoord = world.getHighestBlockYAt(loc) + 1;
+        }
+        boolean found = false;
```

Comment:
What is this loop for? Getting the highest y block should give you a coordinate with enough space for the player, as all blocks above the highest block are air.

Reply:
The y coordinate is randomly generated and may be under the ground (or the player eats underground). Vanilla then selects the next possible location lying under the selected target. So the final target is never above the selected random location.